### Import Libraries


In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

from pathlib import Path

pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", 100)
pd.set_option("display.max_colwidth", 100)

print("Libraries imported successfully.")

Libraries imported successfully.


### Load and verify the raw dataset

In [2]:
# Load and verify the raw dataset

PROJECT_ROOT = Path.cwd().parent
RAW_DATA_PATH = PROJECT_ROOT / "data" / "raw" / "extended_stress_detection_data.csv"

df = pd.read_csv(RAW_DATA_PATH)

print("Dataset loaded successfully.")
print(f"File   : {RAW_DATA_PATH.name}")
print(f"Shape  : {df.shape[0]:,} rows × {df.shape[1]} columns")
print(f"Exists : {RAW_DATA_PATH.exists()}")

Dataset loaded successfully.
File   : extended_stress_detection_data.csv
Shape  : 3,000 rows × 22 columns
Exists : True


In [3]:
# Initial inspection of the raw dataset

display(df.head(10))

print("\nColumn names:")
for i, column in enumerate(df.columns, 1):
    print(f"{i:2}. {column}")

print(f"\nDataset shape: {df.shape[0]:,} rows × {df.shape[1]} columns")

,Age,Gender,Occupation,Marital_Status,Sleep_Duration,Sleep_Quality,Wake_Up_Time,Bed_Time,Physical_Activity,Screen_Time,Caffeine_Intake,Alcohol_Intake,Smoking_Habit,Work_Hours,Travel_Time,Social_Interactions,Meditation_Practice,Exercise_Type,Blood_Pressure,Cholesterol_Level,Blood_Sugar_Level,Stress_Detection
0,30.0,Male,Software Engineer,Single,7.0,4.0,7:00 AM,10:00 PM,2.0,4.0,1,0,No,8.0,1.0,5.0,Yes,Cardio,120.0,180.0,90.0,Low
1,35.0,Female,Marketing Manager,Married,6.0,3.0,6:00 AM,11:00 PM,1.0,3.0,0,1,No,9.0,2.0,3.0,No,Yoga,110.0,160.0,80.0,Medium
2,40.0,Male,Data Scientist,Divorced,7.0,4.0,7:00 AM,10:00 PM,2.0,4.0,1,0,No,8.0,1.0,5.0,Yes,Strength Training,130.0,200.0,100.0,High
3,35.0,Male,Software Engineer,Single,7.0,4.0,7:00 AM,10:00 PM,2.0,4.0,1,0,No,8.0,1.0,5.0,Yes,Cardio,120.0,180.0,90.0,Low
4,29.0,Female,Teacher,Single,8.0,5.0,6:30 AM,10:30 PM,3.0,2.0,1,0,No,7.0,1.0,4.0,Yes,Yoga,110.0,180.0,90.0,Low
5,45.0,Male,Doctor,Married,6.0,3.0,5:30 AM,11:00 PM,2.0,4.0,2,1,Yes,10.0,2.0,3.0,No,Cardio,130.0,220.0,110.0,High
6,32.0,Female,Graphic Designer,Single,7.0,4.0,7:00 AM,12:00 AM,1.0,6.0,3,0,No,9.0,0.5,2.0,No,Aerobics,120.0,190.0,85.0,Medium
7,37.0,Male,Civil Engineer,Married,7.5,4.0,6:00 AM,10:30 PM,2.5,3.5,2,0,No,8.0,1.5,5.0,Yes,Strength Training,140.0,210.0,100.0,Medium
8,50.0,Male,Business Owner,Married,5.5,2.0,5:00 AM,12:30 AM,2.0,8.0,3,2,Yes,12.0,2.0,2.0,No,Walking,150.0,240.0,130.0,High
9,40.0,Female,Nurse,Divorced,6.0,3.0,6:00 AM,11:30 PM,3.0,4.5,2,0,No,9.0,1.0,4.0,Yes,Pilates,125.0,200.0,105.0,Medium



Column names:
 1. Age
 2. Gender
 3. Occupation
 4. Marital_Status
 5. Sleep_Duration
 6. Sleep_Quality
 7. Wake_Up_Time
 8. Bed_Time
 9. Physical_Activity
10. Screen_Time
11. Caffeine_Intake
12. Alcohol_Intake
13. Smoking_Habit
14. Work_Hours
15. Travel_Time
16. Social_Interactions
17. Meditation_Practice
18. Exercise_Type
19. Blood_Pressure
20. Cholesterol_Level
21. Blood_Sugar_Level
22. Stress_Detection

Dataset shape: 3,000 rows × 22 columns


In [4]:
# Data quality audit

audit = pd.DataFrame({
    "Data Type": df.dtypes.astype(str),
    "Non-Null": df.notna().sum(),
    "Missing": df.isna().sum(),
    "Unique Values": df.nunique(dropna=True)
})

print(f"Total records: {len(df):,}")
print(f"Total features: {df.shape[1]}")
print(f"Duplicate rows: {df.duplicated().sum():,}")

display(audit)

Total records: 3,000
Total features: 22
Duplicate rows: 0


,Data Type,Non-Null,Missing,Unique Values
Age,float64,3000,0,1549
Gender,str,3000,0,2
Occupation,str,3000,0,169
Marital_Status,str,3000,0,3
Sleep_Duration,float64,3000,0,549
Sleep_Quality,float64,3000,0,501
Wake_Up_Time,str,3000,0,10
Bed_Time,str,3000,0,19
Physical_Activity,float64,3000,0,560
Screen_Time,float64,3000,0,556


### Feature and Target Classification

In [5]:
# Classify features and target

TARGET = "Stress_Detection"

numeric_features = df.select_dtypes(include=np.number).columns.tolist()
categorical_features = df.select_dtypes(exclude=np.number).columns.tolist()
categorical_features.remove(TARGET)

print(f"Target variable: {TARGET}")
print(f"Input features: {df.shape[1] - 1}")
print(f"Numerical features: {len(numeric_features)}")
print(f"Categorical features: {len(categorical_features)}")

print("\nNumerical features:")
print(numeric_features)

print("\nCategorical features:")
print(categorical_features)

Target variable: Stress_Detection
Input features: 21
Numerical features: 13
Categorical features: 8

Numerical features:
['Age', 'Sleep_Duration', 'Sleep_Quality', 'Physical_Activity', 'Screen_Time', 'Caffeine_Intake', 'Alcohol_Intake', 'Work_Hours', 'Travel_Time', 'Social_Interactions', 'Blood_Pressure', 'Cholesterol_Level', 'Blood_Sugar_Level']

Categorical features:
['Gender', 'Occupation', 'Marital_Status', 'Wake_Up_Time', 'Bed_Time', 'Smoking_Habit', 'Meditation_Practice', 'Exercise_Type']


In [6]:
# Target class distribution

target_summary = (
    df[TARGET]
    .value_counts()
    .rename_axis("Stress Level")
    .to_frame("Count")
)

target_summary["Percentage"] = (
    target_summary["Count"] / len(df) * 100
).round(2)

display(target_summary)

print(f"Number of classes: {df[TARGET].nunique()}")
print(f"Classes: {sorted(df[TARGET].unique())}")

,Count,Percentage
Stress Level,,
Medium,1258,41.93
High,1118,37.27
Low,624,20.80


Number of classes: 3
Classes: ['High', 'Low', 'Medium']


In [7]:
# Numerical and categorical feature summary

print("NUMERICAL FEATURES")
display(df[numeric_features].describe().T.round(2))

print("\nCATEGORICAL FEATURES")
categorical_summary = pd.DataFrame({
    "Unique Values": df[categorical_features].nunique(),
    "Most Frequent": df[categorical_features].mode().iloc[0],
    "Frequency": df[categorical_features].apply(lambda col: col.value_counts().iloc[0])
})

display(categorical_summary)

NUMERICAL FEATURES


,count,mean,std,min,25%,50%,75%,max
Age,3000.0,38.85,7.59,16.98,33.10,38.74,44.47,60.84
Sleep_Duration,3000.0,6.32,1.14,1.92,5.70,6.30,7.00,10.16
Sleep_Quality,3000.0,3.87,0.99,0.52,3.22,3.90,4.43,7.73
Physical_Activity,3000.0,2.96,1.18,-1.34,2.10,3.00,3.85,7.21
Screen_Time,3000.0,4.06,1.17,-0.44,3.21,4.00,5.00,8.64
Caffeine_Intake,3000.0,1.81,0.83,0.00,1.00,2.00,2.00,4.00
Alcohol_Intake,3000.0,0.88,0.66,0.00,0.00,1.00,1.00,2.00
Work_Hours,3000.0,8.25,1.35,3.28,7.31,8.05,9.00,14.00
Travel_Time,3000.0,2.84,1.37,-1.67,1.96,2.97,4.00,7.27
Social_Interactions,3000.0,3.18,1.22,-0.82,2.26,3.04,4.00,7.50



CATEGORICAL FEATURES


,Unique Values,Most Frequent,Frequency
Gender,2,Female,1521
Occupation,169,Teacher,140
Marital_Status,3,Single,1372
Wake_Up_Time,10,7:00 AM,726
Bed_Time,19,6:00 PM,571
Smoking_Habit,2,Yes,1601
Meditation_Practice,2,Yes,1837
Exercise_Type,7,Strength Training,979


## Raw Dataset Audit — Summary

The raw Human Stress Detection Extended Dataset contains **3,000 records and 22 columns**, consisting of **21 input features and one target variable (`Stress_Detection`)**.

The input features include **13 numerical variables** and **8 categorical variables** covering demographic, lifestyle, sleep, occupational, behavioral, and physiological characteristics.

The target variable, `Stress_Detection`, contains three classes:

- **Medium:** 1,258 records (41.93%)
- **High:** 1,118 records (37.27%)
- **Low:** 624 records (20.80%)

The dataset contains **no missing values** and **no exact duplicate rows**.

The categorical variables include `Gender`, `Occupation`, `Marital_Status`, `Wake_Up_Time`, `Bed_Time`, `Smoking_Habit`, `Meditation_Practice`, and `Exercise_Type`. `Occupation` has a relatively high cardinality with 169 unique categories.

Several numerical variables contain values outside their expected intuitive ranges, including negative values in `Physical_Activity`, `Screen_Time`, `Travel_Time`, and `Social_Interactions`. These values are retained in the raw dataset and will be investigated during the exploratory data analysis stage.

The target distribution shows moderate class imbalance, with the `Low` stress class representing the smallest group.

### Audit Conclusion

The raw dataset is structurally complete, with no missing or duplicate records. However, class distribution, unusual numerical values, and high-cardinality categorical variables require further investigation during EDA before preprocessing and graph construction.